From: Joan Torres <joan.torres@suse.com>
Date: Wed, 26 Feb 2025 17:33:08 +0100
Subject: clutter/color-state-params: Handle on gamma EOTFs negative vals

This is a requirement from the wayland color management:

"Negative values are handled by mirroring the positive half of
the curve through the origin."

Forwarded: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4304
Applied-upstream: 48.0, commit:62df470837b81fd89fddf5f7e83272757ba5b7d2
---
 clutter/clutter/clutter-color-state-params.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/clutter/clutter/clutter-color-state-params.c b/clutter/clutter/clutter-color-state-params.c
index 3a67b73..bbaa4b0 100644
--- a/clutter/clutter/clutter-color-state-params.c
+++ b/clutter/clutter/clutter-color-state-params.c
@@ -647,7 +647,10 @@ static const char gamma_eotf_source[] =
   "// Returns: tristimulus values ([0,1])\n"
   "vec3 gamma_eotf (vec3 color)\n"
   "{\n"
-  "  return pow (color, vec3 (" UNIFORM_NAME_GAMMA_EXP "));\n"
+  "  bvec3 is_negative = lessThan (color, vec3 (0.0));"
+  "  vec3 positive = pow (abs (color), vec3 (" UNIFORM_NAME_GAMMA_EXP "));\n"
+  "  vec3 negative = -positive;\n"
+  "  return mix (positive, negative, is_negative);\n"
   "}\n"
   "\n"
   "vec4 gamma_eotf (vec4 color)\n"
@@ -662,7 +665,10 @@ static const char gamma_inv_eotf_source[] =
   "// Returns: Normalized ([0,1]) electrical signal value\n"
   "vec3 gamma_inv_eotf (vec3 color)\n"
   "{\n"
-  "  return pow (color, vec3 (" UNIFORM_NAME_INV_GAMMA_EXP "));\n"
+  "  bvec3 is_negative = lessThan (color, vec3 (0.0));"
+  "  vec3 positive = pow (abs (color), vec3 (" UNIFORM_NAME_INV_GAMMA_EXP "));\n"
+  "  vec3 negative = -positive;\n"
+  "  return mix (positive, negative, is_negative);\n"
   "}\n"
   "\n"
   "vec4 gamma_inv_eotf (vec4 color)\n"
